### PR TITLE
[GPU] Use CuDnnThunk for FMHA.

### PR DESCRIPTION
**Pull request by sergachev**
CuDnnThunk currently used for GEMM fusions is capable of executing arbitrary cuDNN graphs. Moving FMHA to use it lets remove lots of specialized runtime code.

The overview of the change is:
 - cuda_dnn.cc: At cuDNN graph construction assign tensor UIDs using their order in HLO to match the CuDnnThunk calling convention instead of using custom constants.
 - cuda_dnn.h/cc: Move dropout seed / offset / increment to the CudnnGraph properties and handle them accordingly during graph execution.
 - Rename cudnn_workspace_rewriter to cudnn_custom_call_compiler and let it set workspace as it did before + compile and serialize graphs just like cudnn_fusion_compiler aiming CuDnnThunks already does.
 - Move the remainders of the MHA config / descriptor logic to cudnn_custom_call_compiler from the deleted fused_mha_runner.
 - ir_emitter_unnested.cc: Remove MHA-specific logic, create CuDnnThunks for MHA custom calls the same universal way that works for cuDNN GEMM fusions.
 - Delete no more necessary special thunks, runners, lazy ops, command buffer commands.

### CODE REVIEW

**Comment by sergachev** [comment]
I already considered making the change smaller and am still open to ideas how to do that. I hope the description helps navigate it during review.

The MHA config / descriptor logic moved to cudnn_custom_call_compiler can probably be simplified and mostly removed later, I did not do that to avoid further complicating the change.

The code is still covered by xla/service/gpu/tests:gpu_fused_mha_test_gpu.

**Comment by sergachev** [comment]
clang-format check is unhappy because of deleted files, I didn't find any actual formatting problems.

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 1]
```diff
@@ -5083,37 +5058,40 @@ absl::StatusOr<CudnnGraph> GetCudnnFlashAttentionOperationGraph(
...
     sdpa_options.set_padding_mask(true);
     sdpa_options.set_seq_len_q(seq_q_tensor);
     sdpa_options.set_seq_len_kv(seq_kv_tensor);
   }
   // Setting seed and offset
+  std::shared_ptr<Tensor_attributes> seed_tensor;
+  std::shared_ptr<Tensor_attributes> offset_tensor;
   if (use_dropout) {
-    auto seed_tensor =
+    // Skip setting UIDs: pass by value tensors go at the end.
```
Comment (ezhulenev) [inline xla/stream_executor/cuda/cuda_dnn.cc:5078]: Hm, comment seems out of sync, you set uid here and below

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 2]
```diff
@@ -5032,26 +5005,28 @@ absl::StatusOr<CudnnGraph> GetCudnnFlashAttentionOperationGraph(
       .set_io_data_type(ioDataType)
       .set_compute_data_type(cudnn_frontend::DataType_t::FLOAT);
 
+  int uid_offset = 0;
```
Comment (ezhulenev) [inline xla/stream_executor/cuda/cuda_dnn.cc:5008]: Can you maybe add a small doc explaining what is `uid`, without going to cudnn documentation it's hard to guess. If it's obvious for anyone familiar with cudnn, ok to skip it.

Also maybe

```
auto next_uid = [uid = 0]() mutable -> CuDnnTensorUID {
  return CuDnnTensorUID(uid++);
};
```

which makes it very clear how uid is assigned

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 3]
```diff
@@ -5195,71 +5179,66 @@ absl::StatusOr<CudnnGraph> GetCudnnFlashAttentionBackwardOperationGraph(
...
+  bool is_causal = mask_type == dnn::FMHAMaskKind::CAUSAL ||
+                   mask_type == dnn::FMHAMaskKind::PADDING_CAUSAL;
+  auto sdpa_backward_options =
+      cudnn_frontend::graph::SDPA_backward_attributes()
+          .set_name("flash_attention_backward")
+          .set_causal_mask(is_causal)
+          .set_attn_scale(scale)
+          .set_compute_data_type(cudnn_frontend::DataType_t::FLOAT);
+
+  int uid_offset = 0;
```
Comment (ezhulenev) [inline xla/stream_executor/cuda/cuda_dnn.cc:5204]: ditto

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 2 above)
Comment (sergachev) [reply xla/stream_executor/cuda/cuda_dnn.cc]: Documented UIDs and assignment scheme here https://github.com/openxla/xla/pull/15919/files#diff-92794789634f5ed143bb6f39b59c9e8d9c6d15eaf54cb0c6d19859f673dca6a0R32

**Review by ezhulenev — APPROVED** [review]

**Pull request closed without merging**